content/doc/book/pipeline/pipeline-best-practices.adoc

```diff
@@ -0,0 +1,97 @@
+
+Solution: The best solution would be to work around the calls being made but if they must be done then it would be better to implement a Jenkins plugin which is able to gather the data needed.
+
+== Shared Libraries
+
+=== Just say no to overriding built-in Pipeline Steps
+
+Wherever possible stay away from customized/overwritten Pipeline steps. 
+This is the process where the standard Pipeline APIs like `sh` or `timeout` are overwritten using shared libraries. 
+It is dangerous to do this because the Pipeline APIs can change at any time causing custom code to break or give different results than expected. 
+These changes can also become difficult to troubleshoot when something is breaking or causing issues. 
+So even if custom code has not changed that does not mean that after an API update it will keep working the same. 
+Lastly, because of the ubiquitous use of these steps throughout Pipelines, if something is coded incorrectly/inefficiently the results can be catastrophic to Jenkins.
+
+=== Avoid large global variable declaration files
+
+Having large variable declaration file will load every variable for every Pipeline and depending on concurrency can end up using large amounts of memory for very little to no benefit. 
+It would be better to load variables which are relevant to the current execution.
+
+=== Avoid very large shared libraries
+
+This will load the same shared library per-job that is currently executing which can lead to large memory overhead. 
+This can also cause a slow down of every Pipeline since it will need to checkout a very large file every time before starting.
+
+== Additional FAQs
+
+=== How do I deal with Concurrency in Pipelines?
+
+Try not to share workspaces across multiple Pipeline executions or multiple distinct Pipelines. 
+This can lead to either unexpected file modification within each Pipeline or each workspace will get renamed unexpectedly
+
+Ideally shared volumes/disks are mounted somewhere else and the files are copied from one place to the current workspace. 
+Then when the build is done it can be copied back
+
+Build in distinct containers which create needed resources from scratch(cloud-type agents work great for this). 
+This will ensure that the build process happens from scratch every time and is easily repeatable. 
```

Comment:
"This" == Building these containers will ensure that the build process begins at the start every time and is easily repeatable.